lib/auth/keystore/hsm.go

```diff
@@ -0,0 +1,286 @@
+package keystore
+
+import (
+	"crypto"
+	"encoding/json"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/sshutils"
+	"github.com/gravitational/trace"
+
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/google/uuid"
+)
+
+var label = []byte("teleport")
+
+// HSMConfig is used to pass HSM client configuration parameters.
+type HSMConfig struct {
+	// Path is the path to the PKCS11 module.
+	Path string
+	// SlotNumber is the PKCS11 slot to use.
+	SlotNumber *int
+	// TokenLabel is the label of the PKCS11 token to use.
+	TokenLabel string
+	// Pin is the PKCS11 pin for the given token.
+	Pin string
+
+	// HostUUID is the UUID of the local auth server this HSM is connected to.
+	HostUUID string
+}
+
+type hsmKeyStore struct {
+	ctx      *crypto11.Context
+	hostUUID string
+}
+
+func NewHSMKeyStore(config *HSMConfig) (KeyStore, error) {
+	cryptoConfig := &crypto11.Config{
+		Path:       config.Path,
+		TokenLabel: config.TokenLabel,
+		SlotNumber: config.SlotNumber,
+		Pin:        config.Pin,
+	}
+	ctx, err := crypto11.Configure(cryptoConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &hsmKeyStore{
+		ctx:      ctx,
+		hostUUID: config.HostUUID,
+	}, nil
+}
+
+// GenerateRSA creates a new RSA private key and returns its identifier and a
+// crypto.Signer. The returned identifier can be passed to GetSigner later to
+// get the same crypto.Signer.
+func (c *hsmKeyStore) GenerateRSA() ([]byte, crypto.Signer, error) {
+	var id uuid.UUID
+	var signer crypto.Signer
+	var err error
+
+	// Some HSMs (like YubiHSM2) will truncate the passed ID to as few as 2
+	// bytes. There's not a great way to detect this and I don't want to limit
+	// the ID to 2 bytes on all systems, so for now we will generate a few
+	// random IDs and hope to avoid a collision. Ideally Teleport should be the
+	// only thing creating keys for this token and there should only be 10 keys
+	// per HSM at a given time:
+	// 2(rotation phases) * (4(SSH and TLS for User and Host CA) + 1(JWT CA))
+	for iterations := 0; iterations < 16 && (err != nil || signer == nil); iterations++ {
+		id, err = uuid.NewRandom()
+		if err != nil {
+			return nil, nil, err
+		}
+		signer, err = c.ctx.GenerateRSAKeyPairWithLabel(id[:], label, teleport.RSAKeySize)
```

Comment:
So I'm assuming that `GenerateRSAKeyPairWithLabel()` will fail if there is an ID collision - is that true? Is it worth putting in some sort of test to verify this?

Reply:
That's what I was assuming as well, good catch because actually SoftHSM actually will *not* return any error for a duplicate ID (it does let us use the full 16 byte UUID though so we shouldn't get any collisions). YubiHSM does return an error, I haven't tried CloudHSM yet. I'm going to think a bit more about what to do for generating IDs because I don't really like what I have here

Reply:
I've updated this function to first check if the (possibly silently truncated) ID is already taken before generating the keypair